RecoBTag/ONNXRuntime/plugins/ParticleTransformerAK4SonicJetTagsProducer.cc

```diff
@@ -0,0 +1,243 @@
+  for (unsigned jet_n = 0; jet_n < tag_infos->size(); ++jet_n) {
+    max_n_cpf_counter = std::max(max_n_cpf_counter,
+                         static_cast<unsigned int>(((*tag_infos)[jet_n]).features().c_pf_features.size()));
+    max_n_npf_counter = std::max(max_n_npf_counter,
+                         static_cast<unsigned int>(((*tag_infos)[jet_n]).features().n_pf_features.size()));
+    max_n_vtx_counter = std::max(max_n_vtx_counter,
+                         static_cast<unsigned int>(((*tag_infos)[jet_n]).features().sv_features.size()));
+  }
+  
+  // If an event has no jet, or all jets has zero n_cpf, n_npf and n_vtx, the inference is skipped.
+  if (max_n_cpf_counter == 0 && max_n_npf_counter == 0 && max_n_vtx_counter == 0) {
+    client_->setBatchSize(0);
+    skippedInference_ = true;
+    return;
+  }
+    
+  // all the jets in the same event will fill up the same amount of n_cpf, n_npf, n_vtx and send to server
+  const unsigned int max_n_cpf = std::clamp(max_n_cpf_counter, (unsigned int)0, (unsigned int)25);
+  const unsigned int max_n_npf = std::clamp(max_n_npf_counter, (unsigned int)0, (unsigned int)25);
+  const unsigned int max_n_vtx = std::clamp(max_n_vtx_counter, (unsigned int)0, (unsigned int)5);
+
+  for (unsigned igroup = 0; igroup < input_names_.size(); ++igroup) {
+    const auto &group_name = input_names_[igroup];
+    auto &input = iInput.at(group_name);
+    unsigned target = 0;
+    
+    if (igroup == parT::kChargedCandidates || igroup == parT::kChargedCandidates4Vec) target = std::max((unsigned int)1, max_n_cpf);
+    else if (igroup == parT::kNeutralCandidates || igroup == parT::kNeutralCandidates4Vec) target = std::max((unsigned int)1, max_n_npf);
+    else if (igroup == parT::kVertices || igroup == parT::kVertices4Vec) target = std::max((unsigned int)1, max_n_vtx);
+    
+    input.setShape(0, target);
+    auto tdata = input.allocate<float>(true);
+    for (unsigned jet_n = 0; jet_n < tag_infos->size(); ++jet_n) {
+      const auto &taginfo = (*tag_infos)[jet_n];
+      const auto &features = taginfo.features();
+      auto &vdata = (*tdata)[jet_n];
+      // Loop through the n cpf, and in the case that n_cpf is smaller than max_n_cpf, add padding values to all features
+      if (igroup == parT::kChargedCandidates) {
+        unsigned int n_cpf = features.c_pf_features.size();
+        n_cpf = std::clamp(n_cpf, (unsigned int)0,  (unsigned int)25);
+        for (unsigned int count = 0; count < n_cpf; count++) {
+          parT_tensor_filler(vdata, parT::kChargedCandidates, features.c_pf_features.at(count));
+        }
+        if (padding_ && n_cpf < max_n_cpf)
+          vdata.insert(vdata.end(), (max_n_cpf - n_cpf) * parT::n_features_cpf, 0); // Add 0 to unfilled part as padding value
+        if (max_n_cpf == 0) 
+          vdata.insert(vdata.end(), parT::n_features_cpf, 0); // Add at least 1 row of 0 for a jet that has 0 cpf/npf/sv.
+      }
+      else if (igroup == parT::kNeutralCandidates) {
```

Comment:
there's a fair amount of duplicated structure in these successive blocks. you could try to introduce a function that would perform this operation given the necessary inputs for each case: `igroup`, `features.x`, max value for clamp, `max_n_x`, etc.

Reply:
I can do that, should I do similar things to the direct inference code?

Reply:
If possible, yes

Reply:
code changed.